### PR TITLE
Atom/SST and SOF coexistence for ACPI devices

### PR DESCRIPTION
**Pull request by plbossart**
Compile-tested draft to make the selection between Atom/SST and SOF drivers at run-time based on the existing  intel-dsp-config mechanism used for PCI.

Not tested on any hardware for now for lack of time, but feedback welcome.

### CODE REVIEW

File: sound/hda/intel-dsp-config.c
Comment (lyakh) [inline sound/hda/intel-dsp-config.c:526]: you will never get here? You'll return 2 lines above in this case. Maybe you wanted to swap these two ifs and use an `else if` for the `return` path?

File: sound/hda/intel-dsp-config.c
Comment (lyakh) [inline sound/hda/intel-dsp-config.c:507]: `size_t` or something similar

File: sound/hda/intel-dsp-config.c
Comment (lyakh) [inline sound/hda/intel-dsp-config.c]: `memcmp()`?

File: sound/hda/intel-dsp-config.c
Comment (lyakh) [inline sound/hda/intel-dsp-config.c:505]: personally I find passing arrays as parameters (in C) somewhat confusing. In general parameters are passed by value in C, so, the above kind of makes you think, that `ACPI_ID_LEN` values will be passed to the function and changing them in the function will have no effect on the original array. However, that isn't the case, effectively just a pointer is passed. It's valid C and it isn't prohibited by the kernel coding style, so, it's fine, but maybe it'd be clearer to just pass a pointer. Maybe defining a type or a structure would help.

File: sound/soc/intel/boards/bytcht_cx2072x.c
Comment (lyakh) [inline sound/soc/intel/boards/bytcht_cx2072x.c:280]: I found one (!) more example of assigning driver PM ops at runtime in drivers/net/wireless/ti/wlcore/main.c so, presumably, this should work...

File: sound/soc/sof/intel/Kconfig
Comment (lyakh) [inline sound/soc/sof/intel/Kconfig]: is this how we refer to them - "userspace firmware" and "userspace topology?"

**Review by lyakh — CHANGES_REQUESTED** [review]

File: sound/soc/sof/intel/Kconfig
Comment (kv2019i) [inline sound/soc/sof/intel/Kconfig]: typo: coexit->coexist

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c:507]: I want to keep the symmetry with existing code for PCI devices. Jaroslav wrote the code so it must be ok?

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c:505]: I don't know, the parameter comes directly from the mod_devicetable.h definitions. I found it clearer to pass it as is.

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c]: yes, sounds good. thanks for the suggestion

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c:526]: good catch, the two tests should be inverted.

File: sound/soc/intel/boards/bytcht_cx2072x.c
Comment (plbossart) [reply sound/soc/intel/boards/bytcht_cx2072x.c:280]: yes it's a bit hairy but Ranjani and I didn't see any other way to do it.

File: sound/soc/sof/intel/Kconfig
Comment (plbossart) [reply sound/soc/sof/intel/Kconfig]: the previous wording has been there for a while. I am open to suggestions to improve, but all these files are part of the 'we don't break userspace' set of data stored on the filesystem that need to work with a new kernel.

**Comment by plbossart** [comment]
Updated with previous feedback fixed + Broadwell support.

The Kconfig needed can be found here: https://github.com/thesofproject/kconfig/pull/42

File: sound/soc/sof/intel/Kconfig
Comment (plbossart) [reply sound/soc/sof/intel/Kconfig]: fixed, thanks

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c]: fixed, thanks

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c:526]: fixed, thanks

**Comment by plbossart** [comment]
@jwrdegoede @perexg this PR seems to work fine on Baytrail/CherryTrail/Broadwell in my tests, and allows the .config to be simplified quite a bit. Switching drivers can be done with options snd_intel_dspcfg dsp_driver = 2 (SST) or 3 (SOF).

@caseorum this may also be of interest to you for your Broadwell tests, you can switch drivers and compare without having to rebuild.

File: sound/soc/sof/intel/Kconfig
Comment (plbossart) [reply sound/soc/sof/intel/Kconfig]: I tweaked the wording a bit, let me know if it's still unclear.

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c:505]: no change

**Review by ranj063 — DISMISSED** [review]
LGTM

File: sound/hda/intel-dsp-config.c
Comment (lyakh) [inline sound/hda/intel-dsp-config.c:510]: this isn't changing anything really, but actually it should be something like `sizeof(acpi_hid)` now

File: sound/hda/intel-dsp-config.c
Comment (lyakh) [reply sound/hda/intel-dsp-config.c:507]: Sorry, with all the due respect, I comment the code, not the authors :-) I don't know exactly why Jaroslav has chosen that type, but calling a function as `f(ARRAY_SIZE(x))` and then using that argument as a loop counter in the function and declaring it as `u32` seems strange to me.

File: include/sound/soc-acpi.h
Comment (lyakh) [inline include/sound/soc-acpi.h]: sorry for not mentioning this yesterday, it did feel a bit uncomfortable, but it didn't quite crystallise to a specific idea. How about having an enum field instead with values from the `SND_INTEL_DSP_DRIVER_*` set? You could even give that enum a name and declare this variable here with that type. `sof_parent_device` sounds a bit too specific to me.

File: sound/soc/intel/boards/bytcht_cx2072x.c
Comment (lyakh) [reply sound/soc/intel/boards/bytcht_cx2072x.c:280]: Ok, as long as there's only *one* instance of this and other similar devices possible on the system, that should be ok, I guess...

File: sound/hda/intel-dsp-config.c
Comment (lyakh) [inline sound/hda/intel-dsp-config.c:495]: not actually proposing to do this, but just out of curiosity - the above could also have been written as
```
#if IS_ENABLED(CONFIG_SND_SOC_INTEL_CATPT) || IS_ENABLED(CONFIG_SND_SOC_SOF_BROADWELL)
	{
		.flags = 0
#if IS_ENABLED(CONFIG_SND_SOC_SOF_BROADWELL)
			| FLAG_SOF
#endif
#if IS_ENABLED(CONFIG_SND_SOC_INTEL_CATPT)
			| FLAG_SST
#endif
			,
		.acpi_hid = "INT3438"
	},
#endif
```
this would be too ugly even for me to propose, even though it would save one entry in the array :-)

**Review by lyakh — COMMENTED** [review]
None of the comments are critical this time, I leave it to the author's judgement whether or not to address them

**Comment by jwrdegoede** [comment]
I have not looked at the code (no time, sorry) but I like the idea/concept of this very much, so +1 from me and thank you for doing this.

File: sound/hda/intel-dsp-config.c
Comment (perexg) [inline sound/hda/intel-dsp-config.c]: Copy-and-paste typo here: `MODULE_PARM_DESC(acpi_dsp_driver,`

**Comment by perexg** [comment]
> I have not looked at the code (no time, sorry) but I like the idea/concept of this very much, so +1 from me and thank you for doing this.

I join to this. It follows my original idea for the PCI probe.

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c]: Thanks for catching this @perexg. I actually want to use a single parameter, and ignore the 'legacy' case for ACPI. Will remove this.

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c:495]: it's easier to manage if we deal with priority order at the higher level with clearer large blocks.

File: include/sound/soc-acpi.h
Comment (plbossart) [reply include/sound/soc-acpi.h]: well the two cases are SOF or non-SOF. If there was a third option we'd use an enum but here it's not really necessary.

I also don't really like this too much, but I couldn't find a simple way to find what the parent device is. We can brainstorm a bit more on this, I agree this could be nicer.

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c:510]: sure, will change.

File: include/sound/soc-acpi.h
Comment (kv2019i) [reply include/sound/soc-acpi.h]: @plbossart @lyakh One option is "pdev->dev.parent->driver->name". That will return "sof-audio-pci". I looked into this for the common-hdmi stuff, but did not use this in the end, as SST could add common-hdmi support as well.

Comment (plbossart) [reply include/sound/soc-acpi.h]: it's not bad @kv2019i. I only need this for ACPI cases, so it'd be useless indeed for the majority of PCI-based devices. We could have a static inline helper in this file so that there's only one place where we encode the driver name.

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c]: fixed

File: sound/hda/intel-dsp-config.c
Comment (plbossart) [reply sound/hda/intel-dsp-config.c:510]: fixed

Comment (plbossart) [reply sound/hda/intel-dsp-config.c:510]: actually not changed using sizeof generates the following warnings

````
/data/pbossart/ktest/sof-dev/sound/hda/intel-dsp-config.c: In function ‘snd_intel_acpi_dsp_find_config’:
/data/pbossart/ktest/sof-dev/sound/hda/intel-dsp-config.c:510:47: warning: ‘sizeof’ on array function parameter ‘acpi_hid’ will return size of ‘const u8 *’ {aka ‘const unsigned char *’} [-Wsizeof-array-argument]
  510 |   if (memcmp(table->acpi_hid, acpi_hid, sizeof(acpi_hid)))
````

Comment (lyakh) [reply sound/hda/intel-dsp-config.c:510]: there you go, it's an array but not quite... :-/ You could use `sizeof(table->acpi_hid)` then...

File: include/sound/soc-acpi.h
Comment (lyakh) [inline include/sound/soc-acpi.h]: you know what I'm going to propose, right? :-)
```
return dev->parent && dev->parent->driver && dev->parent->driver->name &&
	    !strcmp(dev->parent->driver->name, "sof-audio-acpi");
```

File: sound/soc/intel/boards/bytcht_cx2072x.c
Comment (lyakh) [inline sound/soc/intel/boards/bytcht_cx2072x.c:280]: in fact this could just go under the `if` above, you could also avoid adding the `sof_parent` variable in an earlier patch. So that earlier patch would just do `if (snd_soc_acpi_sof_parent(&pdev->dev))` and now it would become
```
	if (snd_soc_acpi_sof_parent(&pdev->dev)) {
		byt_cht_cx2072x_card.name = SOF_CARD_NAME;
		byt_cht_cx2072x_card.driver_name = SOF_DRIVER_NAME;

		pdev->dev.driver->pm = &snd_soc_pm_ops;
	} else {
		byt_cht_cx2072x_card.name = CARD_NAME;
		byt_cht_cx2072x_card.driver_name = DRIVER_NAME;
	}
```

**Review by lyakh — COMMENTED** [review]
Nothing critical again

File: sound/soc/intel/boards/bdw-rt5677.c
Comment (crojewsk) [inline sound/soc/intel/boards/bdw-rt5677.c:456]: Please split this change into a separate patch.

File: include/sound/soc-acpi.h
Comment (plbossart) [reply include/sound/soc-acpi.h]: sounds good.

File: sound/soc/intel/boards/bytcht_cx2072x.c
Comment (plbossart) [reply sound/soc/intel/boards/bytcht_cx2072x.c:280]: i really want to keep the two parts separate, the card name really has nothing to do with power management. It's easier to read and understand if you keep the blocks separate. It's also not critical time-wise, this happens on probe before the card registration.

File: sound/hda/intel-dsp-config.c
Comment (kv2019i) [inline sound/hda/intel-dsp-config.c:390]: I didn't immediately understand the three ids are all Haswell, but yeah, it seems they all are so this is good.

File: sound/soc/intel/boards/bdw-rt5677.c
Comment (kv2019i) [inline sound/soc/intel/boards/bdw-rt5677.c:456]: @plbossart Hmm, I'm a bit confused why here it's ok to set the pm ops without #ifdefs , but in other machines drivers used by both SST and SOF, it's not. I'm probably missing something...

**Review by kv2019i — COMMENTED** [review]
Looks good now.  A couple of question/comments, but nothing blocking the merge.

**Review by crojewsk — COMMENTED** [review]
In general my opinion stands with intel-dsp-config being a testament of chaos existing in Intel part of ASoC. Heavily against any new additions in any form to it (in fact, I'd prefer this module removed). While I understand that it currently allows for granular selection of hda-based devices, in world where there is no hda or where hda co-exists with different devices (separate cards, e.g.: hsw/bdw) there is no need for further complication via intel-dsp-config. In hda world, well, maybe we have to re-think the problem.

So:
ALSA: hda: intel-dsp-config: add helper for ACPI DSP driver selection
ALSA: hda: intel-dsp-config: add Broadwell ACPI DSP driver selection
ASoC: Intel: broadwell: set card and driver name dynamically
ASoC: Intel: catpt: add dynamic selection of DSP driver
ASoC: SOF: Intel: allow for coexistence between SOF and catpt drivers 
<anything I've missed here>

are patches I'm against. Officially, as per business unit decision there is no SOF on hsw/bdw devices, so there should be no code in linux kernel targeting these as effectively that code is rendered redundant. If you do not intend to remove the code, then at least do not complicate other stuff because of that. Separate Kconfigs can be left so user can compile kernel themselves and enable some experimental support if they really insist to do so.

I believe similar decision has been made in regard to byt/cht/bsw. TLDR: either SOF should entirely replace /atom/ -or- /atom/ should be left alone and the only option without anything in intel-dsp-config mentioning it.

For machine board pm changes, as these passed our regression testing:
Tested-by: Cezary Rojewski <cezary.rojewski@intel.com>

File: sound/soc/intel/boards/bdw-rt5677.c
Comment (plbossart) [reply sound/soc/intel/boards/bdw-rt5677.c:456]: The Atom/SST driver has its own power management sequences that I frankly don't understand well, we don't want to interfere with it. For Broadwell, there is no known issue so we can set this unconditionally.

**Comment by plbossart** [comment]
> In general my opinion stands with intel-dsp-config being a testament of chaos existing in Intel part of ASoC. 

Please keep in mind who you work for when writing comments in public.

**Comment by crojewsk** [comment]
Well, several upstream threads already touched on this. Make no mistake - I want my, my team and teams we're cooperating with to represent Intel in the best fashion possible. Don't believe transparency when making crucial design decisions for code-to-be-upstreamed hinders that goal.

**Comment by plbossart** [comment]
> Well, several upstream threads already touched on this. Make no mistake - I want my, my team and teams we're cooperating with to represent Intel in the best fashion possible. Don't believe transparency when making crucial design decisions for code-to-be-upstreamed hinders that goal.

You've got a track record of using unacceptable language offending colleagues and partners, that has to stop.

**Comment by crojewsk** [comment]
I don't understand where this is coming from and what it has to do with the review.

Anyway, do you have any technical arguments for the issues I've risen?

**Review by RanderWang — DISMISSED** [review]
LGTM

**Comment by plbossart** [comment]
re-pushed same code with a different SHA1 to re-trigger the entire CI, now that targets support the dsp_driver config and Kconfigs are updated.
Let's see how this works. Thanks @fredoh9 for your support!

**Comment by plbossart** [comment]
@fredoh9 this doesn't seem to work on ACPI platforms, I think the scripts for alsa-bat assume the card is number 0, and that's no longer true now that I enable HDaudio as well.

**Review by lyakh — APPROVED** [review]

**Review by RanderWang — APPROVED** [review]

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
Underflow reported on Chromebook Cyan (https://sof-ci.01.org/linuxpr/PR2500/build4700/devicetest/), filed as issue https://github.com/thesofproject/sof/issues/3534

This shouldn't have anything to do with this PR, will re-run the tests.

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
@kv2019i @ranj063 @bardliao the SOF CI re-tests show we have an underflow issue that's not caused by this PR.

I'll let you merge if you're ok with the changes.
Thanks!

**Review by ranj063 — APPROVED** [review]

**Comment by xiulipan** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
missing rmmod snd-soc-skl in kmod scripts, corrected now by @xiulipan. let's run one more test and merge, this affects all other PRs

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
I really have no idea what happens on CI devices, so will take the calculated risk of merging so that we can fix remaining CI deployment/sof-test issues. Thanks for your understanding y'all.

**Comment by kv2019i** [comment]
Ack @plbossart . I'm seeing some weird errors on other PRs as well right now like https://sof-ci.01.org/linuxpr/PR2518/build4731/devicetest/

**Comment by marc-hb** [comment]
Collected some similar failures in https://github.com/thesofproject/sof-test/issues/458#issuecomment-713736185

@fredoh9  is looking at them

**Comment by fredoh9** [comment]
https://github.com/thesofproject/linux/pull/2500 is required for BYT/BSW/BDW acpi based platforms. When we build linux PR, it checkout that branch only. We need to ask PR owner to rebase top of topic/sof-dev.
For SOF PR, it uses daily kernel images. So it will take 12 hour+(from now) to take today's build.
For sof-tet PR, it uses daily kernel image + dailiy FW image. This takes some hours too.

**Comment by fredoh9** [comment]
>We need to ask PR owner to rebase top of topic/sof-dev.

I thought for PR build it checkout the PR's branch, but it's not. It merge the PR top of topic/sof-dev. Sorry for misunderstanding.

**Comment by marc-hb** [comment]
> For SOF PR, it uses daily kernel images. So it will take 12 hour+(from now) to take today's build.
> For sof-test PR, it uses daily kernel image + dailiy FW image. This takes some hours too.

I looked at some timestamps and I think I confirmed this in https://github.com/thesofproject/sof-test/issues/458

**Pull request merged**
